src/bin/example-read.rs

```diff
@@ -0,0 +1,151 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+extern crate parquet;
+
+use std::fs::File;
+use std::path::Path;
+
+use parquet::basic::*;
+use parquet::data_type::*;
+use parquet::column::reader::{ColumnReaderImpl, get_typed_column_reader};
+use parquet::file::reader::{FileReader, SerializedFileReader};
+
+// Run example
+// ===
+// $ cargo run --bin example-read
+//
+// Console output
+// ===
+// Reading batch of records: 4
+// Record { b: 2, d_c: "abc", e: [1, 1] }
+// Record { b: 3, d_c: "def", e: [10, 10] }
+// Record { b: 4, d_c: "ghi", e: [100, 100] }
+// Record { b: 4, d_c: "jkl", e: [1000, 1000] }
+//
+
+#[derive(Clone, Debug, Default)]
+struct Record { b: i64, d_c: String, e: Vec<i32> }
+
+// read values from column reader for requested batch size, can read fewer records
+fn read_values<T: DataType>(
+    reader: &mut ColumnReaderImpl<T>,
+    batch_size: usize) -> (usize, Vec<T::T>, Vec<i16>, Vec<i16>) where T: 'static {
+
+  let mut values = vec![T::T::default(); batch_size];
+  let mut def_levels = vec![i16::default(); batch_size];
+  let mut rep_levels = vec![i16::default(); batch_size];
+  let mut curr_values_read = 0;
+  let mut curr_levels_read = 0;
+  loop {
+    let (values_read, levels_read) = reader.read_batch(
+      batch_size,
+      Some(&mut def_levels[curr_levels_read..]),
+      Some(&mut rep_levels[curr_levels_read..]),
+      &mut values[curr_values_read..]
+    ).unwrap();
+
+    curr_values_read += values_read;
+    curr_levels_read += levels_read;
+
+    if values_read == 0 {
+      break;
+    }
+  }
+  (curr_values_read, values, def_levels, rep_levels)
+}
+
+fn main() {
```

Comment:
Similar to `read-metadata`, I got this error:
```
➜  parquet-rs (example) ✗ cargo run --bin example-read
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/example-read`
Reading batch of records: 4
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: NYI("Unsupported encoding PLAIN_DICTIONARY")', src/libcore/result.rs:860:4
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Reply:
Could you try fetching latest commits from master? I think this is one of the problems that we solved recently. It works from my branch that contains those fixes.